src/pkg/cli/client/byoc/aws/domain.go

```diff
@@ -54,44 +57,58 @@ func prepareDomainDelegation(ctx context.Context, projectDomain string, r53Clien
 	return delegationSet.NameServers, delegationSetId, nil
 }
 
-func getOrCreateDelegationSet(ctx context.Context, r53Client aws.Route53API) (*types.DelegationSet, error) {
-	// Avoid creating a new delegation set if one already exists
-	delegationSet, err := aws.GetDelegationSet(ctx, r53Client)
-	// Create a new delegation set if it doesn't exist
-	if errors.Is(err, aws.ErrNoDelegationSetFound) {
-		// Create a new delegation set. There's a race condition here, where two deployments could create two different delegation sets,
-		// but this is acceptable because the next time the zone is deployed, we'll get the existing delegation set from the zone.
-		delegationSet, err = aws.CreateDelegationSet(ctx, nil, r53Client)
-	}
-	if err != nil {
-		return nil, err
+func createUsableDelegationSet(ctx context.Context, domain string, r53Client aws.Route53API, resolverAt func(string) dns.Resolver) (*types.DelegationSet, error) {
+	// Try up to 5 times to create a delegation set that is usable (i.e., none of its NS servers have conflicting records for the domain)
+	// Chances of a conflict happened in a single try if aws have 2000 dns servers is about (1 - (1-4/2000)^4) ~ 0.8%
+	// Chances of this happening in 10 consecutive tries if servers are randomly chosen is about 0.8%^5 ~ 3.2e-13, virtually impossible
+	for range 5 {
+		delegationSet, err := aws.CreateDelegationSet(ctx, nil, r53Client)
+		if err != nil {
+			return nil, err
+		}
+		// Verify that the delegation set is usable by checking if any of its NS servers contains conflicting records
+		conflictFound := false
+		for _, nsServer := range delegationSet.NameServers {
+			resolver := resolverAt(nsServer)
+			records, err := resolver.LookupNS(ctx, domain)
+			if err != nil {
+				return nil, err
+			}
+			if len(records) > 0 {
+				// Records found, meaning the NS server is conflicting
+				term.Debugf("Delegation set NS server %q has conflicting records: %v", nsServer, records)
+				conflictFound = true
+				break
+			}
+		}
+		if conflictFound {
+			if err := aws.DeleteDelegationSet(ctx, delegationSet.Id, r53Client); err != nil {
```

Comment:
Can we add a comment that explains why it's OK to continue if this fails. ie. Will it ever get cleaned up? What happens if we create more delegation sets than we need?

Reply:
added